### PR TITLE
Align transaction fields to transaction reference spec

### PR DESCRIPTION
**Pull request by algochoi**
This PR contains breaking changes in transaction fields that align more closely to the [transaction reference spec](https://developer.algorand.org/docs/get-details/transactions/transactions/).

Changes:
* from -> sender
* to -> receiver
* firstRound -> firstValid
* lastRound -> lastValid
* freezeState -> assetFrozen
* assetRevocationTarget -> assetSender
* reKeyTo -> rekeyTo

https://github.com/algorand/js-algorand-sdk/issues/144

### CODE REVIEW

**Comment by algochoi** [comment]
@algoanne We have two places where we use `application-index` in our APIs: https://developer.algorand.org/docs/rest-apis/algod/#pendingtransactionresponse but the most recent ones refer to it as `application-id`. Do we want to make this change here for app/assets?

**Comment by algoanne** [comment]
@algochoi it seems ID is more common. Disappointing that we're not fully consistent at the API level. I really don't feel strongly one way or the other, they both make me sad. I leave it to you.

**Comment by algochoi** [comment]
They also make me sad, and I would like to consolidate everything to the newer convention of `IDs`. 

However, we also have places in the API that use `Index` such as Assets: https://github.com/algorand/go-algorand/blob/master/daemon/algod/api/algod.oas2.json#L2863
Because of these inconsistencies that will also affect how the SDK types will be generated, e.g. https://github.com/algorand/js-algorand-sdk/blob/develop/src/client/v2/algod/models/types.ts#L1049, I'm inclined to keep everything as-is to be consistent with the API spec. I think changing these index-related conventions should be following a breaking change in the API spec as well.

File: src/types/transactions/base.ts [hunk 1]
```diff
@@ -182,12 +182,12 @@ export interface TransactionParams {
   /**
    * Integer first protocol round on which this txn is valid
    */
-  firstRound: number;
+  firstValid: number;
```
Comment (jasonpaulos) [inline src/types/transactions/base.ts:185]: Can you also rename the `reKeyTo` field to `rekeyTo`?

Comment (algochoi) [reply src/types/transactions/base.ts:185]: good point -- [1258d1a](https://github.com/algorand/js-algorand-sdk/pull/804/commits/1258d1a6864851b6a5816de5bd3428e52171fe4c)

**Comment by jasonpaulos** [comment]
> They also make me sad, and I would like to consolidate everything to the newer convention of `IDs`.
> 
> However, we also have places in the API that use [`Index` such as Assets](https://github.com/algorand/go-algorand/blob/master/daemon/algod/api/algod.oas2.json#L2863). Because of these inconsistencies that will also affect how the SDK types will be generated, [e.g. here](https://github.com/algorand/js-algorand-sdk/blob/develop/src/client/v2/algod/models/types.ts#L1049), I'm inclined to keep everything as-is to be consistent with the API spec. I think changing these index-related conventions should be following a breaking change in the API spec as well.

I have a hard time believing that we'll want to break the REST APIs anytime soon with an index => ID change, so it's worth thinking about what we can do short of that solving the problem.

We do have some cases where the REST API uses ID instead of index, for example [here](https://github.com/algorand/js-algorand-sdk/blob/4e1073d6a15b443a73dedf42b025173b397624b0/src/client/v2/algod/models/types.ts#L1113).

One option we could consider is modifying the generator repo to rename certain fields from index => ID. Then we could use ID everywhere, including in the SDK defined classes.

**Comment by algochoi** [comment]
> I have a hard time believing that we'll want to break the REST APIs anytime soon with an index => ID change, so it's worth thinking about what we can do short of that solving the problem.

I agree as well -- I wanted to skirt around this issue for now because it seemed like the most onerous change (requires some custom overrides in generator, and then it may affect the three SDKs we generate for in order to keep field names consistent with each other, plus any overhead that these changes might have in the tests/examples/dev docs). On the other hand, the current changes are pretty isolated to the JS SDK and are easier wins and has had enough notice to dev-rel.

**Comment by algoanne** [comment]
> I agree as well -- I wanted to skirt around this issue for now because it seemed like the most onerous change (requires some custom overrides in generator, and then it may affect the three SDKs we generate for in order to keep field names consistent with each other, plus any overhead that these changes might have in the tests/examples/dev docs). On the other hand, the current changes are pretty isolated to the JS SDK and are easier wins and has had enough notice to dev-rel.

I think I support breaking out the index -> ID changes into a separate PR, since it potentially has broader scope (across all the SDKs). Although we have not been explicitly planning a 3.0 for the other SDKs, it's certainly on the table. It would be valuable to bring this same consistency to all the SDKs. We can discuss. 
Do the other SDKs have consistent naming with the API in the other fields, e.g. lastValid etc. ?

**Comment by algochoi** [comment]
> Do the other SDKs have consistent naming with the API in the other fields, e.g. lastValid etc. ?

I have only checked the obvious ones listed above (sender, lastValid), but yes. The go-sdk closely follows go-algorand conventions, Java is pretty similar, and Python has some minor variations in naming but still makes sense to me. 

The generated SDKs (Go, Java) are also consistent with the API spec, which has some weird naming like using `index` for Assets but `asset-id` for AssetHoldings. So I think changing these would probably warrant a change in all the SDKs. I haven't thought about potential downstream impacts here (e.g. Swift, C# SDKs...).

**Comment by winder** [comment]
I'm fine with ignoring the ID/Index issue, but we also have a lot of flexibility in the generator with applying things to a single SDK

**Review by Eric-Warehime — APPROVED** [review]

**Review by winder — APPROVED** [review]

**Review by bbroder-algo — APPROVED** [review]
s-tier

**Pull request merged**